rayon-core/src/thread_pool/mod.rs

```diff
@@ -461,6 +461,39 @@ pub fn yield_local() -> Option<Yield> {
     }
 }
 
+/// Waits for termination of the thread-pool (if pending), and cleans up resources allocated by
+/// [`ThreadPoolBuilder::use_current_thread()`]. Should only be called from the thread that built
+/// the thread-pool, and only when [`ThreadPoolBuilder::use_current_thread()`] is used.
+///
+/// Calling this function from a thread pool job will block indefinitely.
+///
+/// Calling this function before before the thread-pool has been dropped will cause the thread to
+/// not return control flow to the caller until that happens (stealing work as necessary).
+///
+/// # Panics
+///
+/// If the calling thread is no the creator thread of a thread-pool, or not part of that
+/// thread-pool, via [`ThreadPoolBuilder::use_current_thread()`].
+pub fn clean_up_use_current_thread() {
```

Comment:
Concerning the name: How about changing the verb to avoid doubling it, e.g. `release_current_thread`, `drop_current_thread`?

More generally, could this become part of `impl Drop for ThreadPool` (automatically detecting if the `use_current_thread` was used) as the thread pool created by `build_global` would never be dropped anyway?

Reply:
> Concerning the name: How about changing the verb to avoid doubling it, e.g. `release_current_thread`, `drop_current_thread`?

I'm fine with either. I kinda prefer `use_current_thread` or similar being in the name, because this is specific to that API. Maybe `stop_using_current_thread`?

> More generally, could this become part of `impl Drop for ThreadPool` (automatically detecting if the `use_current_thread` was used) as the thread pool created by `build_global` would never be dropped anyway?

Not really, because:

 * `Drop for ThreadPool` has no guarantee of running on the same thread as `ThreadPoolBuilder::build`.
 * The external thread is not guaranteed to ever yield to Rayon's pool.
 * `ThreadPool`'s drop impl doesn't guarantee that all of the threads have exited anyways (because it might be running from a worker).

Reply:
Maybe it could be rolled into a `ThreadPool::exit(self)` method, somewhat like #688? So if the "used" thread is the one to call that, we can also handle the extra cleanup. If another thread takes the `ThreadPool` and calls it, it's still *safe* to leak that one.

(Though back then you wanted that for the global pool, and also thread destructors are still hard to promise.)

Reply:
I'm ok rolling `clean_up_use_current_thread` into a `ThreadPool::exit` or probably `impl Drop for ThreadPool`?

Reply:
I sent a follow-up patch doing that [here](https://github.com/rayon-rs/rayon/pull/1063/commits/a6653df1f974eaa8b8a7d99e15763516ef58d59a), let me know if that looks good, it's definitely simpler.

If somebody needs to clean-up dropping the thread-pool in a different thread, sending a patch re-introducing the clean-up API seems fine.

Reply:
(Now with CI passing)